rollup/internal/controller/sender/transaction_signer.go

```diff
@@ -0,0 +1,153 @@
+)
+
+const (
+	// PrivateKeySignerType
+	PrivateKeySignerType = "PrivateKey"
+
+	// RemoteSignerSignerType
+	RemoteSignerSignerType = "RemoteSigner"
+)
+
+// TransactionSigner signs given transactions
+type TransactionSigner struct {
+	config    *config.SignerConfig
+	auth      *bind.TransactOpts
+	rpcClient *rpc.Client
+	nonce     uint64
+	addr      common.Address
+}
+
+func NewTransactionSigner(config *config.SignerConfig, chainID *big.Int) (*TransactionSigner, error) {
+	switch config.SignerType {
+	case PrivateKeySignerType:
+		privKey, err := crypto.ToECDSA(common.FromHex(config.PrivateKey))
+		if err != nil {
+			return nil, fmt.Errorf("parse sender private key failed: %w", err)
+		}
+		auth, err := bind.NewKeyedTransactorWithChainID(privKey, chainID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create transactor with chain ID %v, err: %w", chainID, err)
+		}
+		return &TransactionSigner{
+			config:    config,
+			auth:      auth,
+			rpcClient: nil,
+			addr:      crypto.PubkeyToAddress(privKey.PublicKey),
+		}, nil
+	case RemoteSignerSignerType:
+		if config.SignerAddress == "" {
+			return nil, fmt.Errorf("failed to create RemoteSigner, signer address is empty")
+		}
+		rpcClient, err := rpc.Dial(config.RemoteSignerUrl)
+		if err != nil {
+			return nil, fmt.Errorf("failed to dial rpc client, err: %w", err)
+		}
+		return &TransactionSigner{
+			config:    config,
+			rpcClient: rpcClient,
+			addr:      common.HexToAddress(config.SignerAddress),
+		}, nil
+	default:
+		return nil, fmt.Errorf("failed to create new transaction signer, unknown type: %v", config.SignerType)
+	}
+}
+
+func (ts *TransactionSigner) SignTransaction(ctx context.Context, tx *gethTypes.Transaction) (*gethTypes.Transaction, error) {
+	switch ts.config.SignerType {
+	case PrivateKeySignerType:
+		signedTx, err := ts.auth.Signer(ts.addr, tx)
+		if err != nil {
+			log.Error("failed to sign tx", "address", ts.addr.String(), "err", err)
+			return nil, err
+		}
+		return signedTx, nil
+	case RemoteSignerSignerType:
+		rpcTx, err := txDataToRpcTx(&ts.addr, tx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert txData to rpc transaction, err: %w", err)
+		}
+		var result hexutil.Bytes
+		err = ts.rpcClient.CallContext(ctx, &result, "eth_signTransaction", rpcTx)
+		if err != nil {
+			return nil, err
+		}
+		signedTx := new(gethTypes.Transaction)
+		if err := signedTx.UnmarshalBinary(result); err != nil {
+			return nil, err
+		}
+		return signedTx, nil
+	default:
+		// this shouldn't happen, because SignerType is checked during creation
+		return nil, fmt.Errorf("shouldn't happen, unknown signer type")
+	}
+}
+
+func (ts *TransactionSigner) SetNonce(nonce uint64) {
+	ts.nonce = nonce
+}
+
+func (ts *TransactionSigner) GetNonce() uint64 {
+	return ts.nonce
+}
+
+func (ts *TransactionSigner) GetAddr() common.Address {
+	return ts.addr
+}
+
+// RpcTransaction transaction that will be send through rpc to web3Signer
+type RpcTransaction struct {
```

Comment:
Would this structure lose access lists when sending txns with access list? currently, the sender uses access lists to save a bit of cost.

Reply:
e.g. 1000 gas for this commit txn: https://etherscan.io/tx/0xeff1fd655caa507299ed72e4f51b7eabc300e20c6c2e54f2aa80ac2996a5a35f#accesslist

Reply:
yes, access lists are not supported by `web3signer`, unfortunately

Reply:
got it. no problem IMO. but you probably should review the get gas limit logic and make it compatible with `web3signer`, because with access lists, the gas limit would be a little less than without access lists.

one naive method I can think of: when using web3signer, don't try adding access lists when constructing the tx.